### PR TITLE
refactor indexset: support for index mapping from any type to comparable one

### PR DESCRIPTION
**Pull request by MartinForReal**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
refactor indexset: support for index mapping from any type to comparable one
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/provider/loadbalancer/fnutil/slice.go [hunk 1]
```diff
@@ -53,21 +53,55 @@ func IsAll[T any](f func(T) bool, xs []T) bool {
...
+func IndexSet[DATA comparable](xs []DATA) *IndexSetWithComparableIndex[DATA, DATA] {
+	return NewIndexSetWithComparableIndex(PlanHashCode, xs)
+}
+
+type IndexSetWithComparableIndex[INDEX comparable, DATA any] struct {
+	hashCode func(data DATA) INDEX
+	data     map[INDEX]DATA
+}
+
+func NewIndexSetWithComparableIndex[INDEX comparable, DATA any](hashCode func(data DATA) INDEX, xs []DATA) *IndexSetWithComparableIndex[INDEX, DATA] {
```
Comment (zarvd) [inline pkg/provider/loadbalancer/fnutil/slice.go:67]: nit. I believe a single letter such as `I, T, or E` is clear enough for generic type names. Or `PascalCase` maybe.

Comment (zarvd) [inline pkg/provider/loadbalancer/fnutil/slice.go:67]: The functionality looks good to me. However, would a simple map be sufficient for this case? 🤔 I feel like Go usually leans towards using `map` or `slice` for most of its data structures.

Comment (MartinForReal) [reply pkg/provider/loadbalancer/fnutil/slice.go:67]: The mapping function may differ, and it is easier to keep index consistent. 
IMO, it is implementation details, as long as the signature is simple. It's ok

**Review by zarvd — APPROVED** [review]

**Review by cvvz — APPROVED** [review]

**Pull request merged**